libraries/plugins/multiparty_messaging/multiparty_messaging_plugin.cpp

```diff
@@ -0,0 +1,264 @@
+{
+   return _db.find< group_object, by_current_name >( name );
+}
+
+template<typename T> void multiparty_messaging_plugin_impl::save_message(const group_object& go, const account_name_type sender, bool system_message, const T& data)const
+{
+   _db.create<message_object>([&](message_object& mo){
+        mo.group_name = go.group_name;
+        mo.sequence = go.current_seq;
+        mo.sender = sender;
+        mo.recipients = go.members;
+        mo.system_message = system_message;
+        std::copy( data.begin(), data.end(), std::back_inserter(mo.data));
+   });
+   _db.modify(go, [&]( group_object& go){
+        go.current_seq++;
+   });
+}
+
+fc::sha256 multiparty_messaging_plugin_impl::extract_key( const std::map<public_key_type, encrypted_key>& new_key_map, const fc::sha256& group_key, const fc::sha256& iv, const public_key_type& sender_key) const
+{
+   //first look for shared secret key
+   for( const auto& pk: _self._private_keys ){
+      const auto &nkm_itr =  new_key_map.find(pk.first);
+      if(  nkm_itr != new_key_map.end() ){
+         fc::sha512 sc = pk.second.get_shared_secret(sender_key);
+         vector<char> key_data = fc::aes_decrypt( sc, nkm_itr->second );
+         fc::sha256 key( key_data.data(), key_data.size());
+         return key;
+      }
+      //in case we are the sender, there won't be our key in the map... Let's try this:
+      if( pk.first == sender_key) {
+         auto nkm_itr = new_key_map.begin();
+         while( nkm_itr!= new_key_map.end() && nkm_itr->first == public_key_type())
+            nkm_itr++;
+         if( nkm_itr== new_key_map.end() )
+            continue;
+         fc::sha512 sc = pk.second.get_shared_secret(nkm_itr->first);
+         vector<char> key_data = fc::aes_decrypt(sc, nkm_itr->second);
+         fc::sha256 key(key_data.data(), key_data.size());
+         return key;
+      }
+   }
+
+   const auto &nkm_itr =  new_key_map.find(public_key_type());
+   if(  nkm_itr != new_key_map.end() ){
+      vector<char> key_data = fc::aes_decrypt( group_key, iv, nkm_itr->second );
+      fc::sha256 key ( key_data.data(), key_data.size());
+      return key;
+   }
+
+   return fc::sha256();
+}
+
+void multiparty_messaging_plugin_impl::apply( const protocol::custom_json_operation& op )
+{
+   group_meta message_meta = fc::json::from_string(&op.json[ 0 ]).as<group_meta>();
+
+   if( message_meta.iv ) //message sent to the group
+   {
+      //check to which address this message came, and by whom
+      for(account_name_type r: op.recipients){
+         const group_object* g_ob = find_group(r);
+         if( !g_ob ) continue;
+
+         message_wrapper message_content = decode_message( message_meta.data, *message_meta.iv, g_ob->group_key);
+         if( message_content.type == message_wrapper::message_type::group_operation ){
+            //process_group_operation( *message_content.operation_data);
+            FC_ASSERT(message_content.operation_data);
```

Comment:
Can you add comments to asserts, so when it happens people are not running to us asking for questions what is on line #.. in cpp file ...

Reply:
done